### ISSUE TITLE
Poor performance with command window

### ISSUE THREAD

**Issue by robsmith11** [tridactyl/tridactyl#1092]
I am trying out tridactyl on my Android tablet again and the performance is quite good (hint drawing, tab switching, scrolling, etc.) except for interacting with the command window.

It will take 2 seconds for the window to appear (1 second for the single line bar to appear and then another second for it to fully expand).  And typing in it is very laggy (when I type quickly, characters will be drawn 5-6 key-presses late).

Obviously my CPU (Snapdragon 820) is partly to blame, but I feel like performance could be quite a bit better, expecially for the laggy text input.  

Android 8.1, Firefox Nightly, Tridactyl 1.14

EDIT:
Typing lag is fixed with latest beta of Tridactyl.  Command window opening lag remains.

**Comment by glacambre** [comment]
Does this also happen on the betas?

**Comment by robsmith11** [comment]
Beta Firefox?  Yes, similar performance.

There is a slight difference between the arm32 and arm64 builds with arm64 being slower because the Ion JIT compiler has not been ported yet.  But even with arm32, the performance of the command window is far slower than I would expect.

**Comment by glacambre** [comment]
No, I meant Tridactyl betas, sorry. You can find them here: https://tridactyl.cmcaine.co.uk/betas/?sort=time&order=desc

I expect them to fix the laggy input (although completions will still take a while to update).

**Comment by robsmith11** [comment]
Ah thanks.  With the latest beta, yes, all lag with typing is gone!

The command window, however, still takes a 1-2 seconds to appear and with the beta appears to be a bit buggy.  It looks like most of it is cut off (everything above the bottom line of text) and pressing `Esc` does not close it.

**Comment by glacambre** [comment]
>It looks like most of it is cut off

I'm not sure I understand what you mean, could you upload a screenshot?

**Comment by robsmith11** [comment]
Yeah sorry my description was not very clear.

I'm using a fresh install of Firefox Beta for Android (no addons other than Tridactyl Beta).  In all sreenshots below, I opened the command window by pressing `o`.

The command window opens/closes without any problems when on the Tridactyl tutorial page:
https://github.com/robsmith11/ss/blob/master/Screenshot_Firefox_Beta_20181018-163934.png

But for all other pages the command window only opens partially, covering up the text entry:
https://github.com/robsmith11/ss/blob/master/Screenshot_Firefox_Beta_20181018-164002.png

If I attempt to enter a URL, then the text entry portion become visible, replacing the "History" title, and if I hit `Enter`, it works properly:
https://github.com/robsmith11/ss/blob/master/Screenshot_Firefox_Beta_20181018-164013.png

However, if I attempt to close the command window by pressing `Esc`, the text entry is cleared, but the window is not closed:
https://github.com/robsmith11/ss/blob/master/Screenshot_Firefox_Beta_20181018-164023.png

I am only able to close the window by either pressing `Enter` to execute a command or by shifting focus to another element in the webpage and then pressing `Esc`.

Android does not have history WebExtension API support yet, so that may be part of the problem.  But that should just cause there to be no completions.

**Comment by glacambre** [comment]
Thank you for taking the time to upload these screenshots and for detailing the steps you took, this is very useful.

I implemented a few changes that I feel might help, could you install [this addon](https://github.com/tridactyl/tridactyl/files/2493195/tridactyl-1.14.0.zip) and tell us if it solves any of the issues you encounter? It is built from https://github.com/glacambre/tridactyl/commit/36baf413b83d1122c80019adda8068a953cc2814 if you want to know what it does exactly. In order to install it you'll probably need to set `xpinstall.signatures.required` to false in `about:config`, to download it and then to load it using the "install addon from file" menu in `about:addons`.

**Comment by robsmith11** [comment]
Thanks.  I tested your new version, but unfortunately, I do not see any difference in terms of speed or the behavior in my screenshots.

However, I noticed that with your latest version and more recent betas, the typing lag has returned.
Last fast-typing beta:
https://tridactyl.cmcaine.co.uk/betas/tridactyl-1.14.0pre2013-an+fx.xpi
Typing is slow again (very noticable on my tablet) from this beta forward:
https://tridactyl.cmcaine.co.uk/betas/tridactyl-1.14.0pre2015-an+fx.xpi

**Comment by robsmith11** [comment]
Seems this commit is to blame for the typing speed fix regression:

https://github.com/tridactyl/tridactyl/commit/6a9cab153ccee0dd99bad26f3306406fe72fb737#diff-df4ab8e8e6345960f74e80de5a1419e3L234

**Comment by pickfire** [comment]
The performance is also poor on Firefox desktop. Especially when typing ` ` after quite a few words (~10 words). The search a lot slower when compared to the default Firefox searchbar.

**Comment by Z3po** [comment]
I'm also recognizing since the latest version of firefox a huge input lag. This is really painfull when writing multiple words with spaces. I'm not sure if that's caused by the newest firefox version or from tridactyl itself.
I'm currently using:
* Tridactyl 1.14.8-90-g4c2a702
* firefox-65.0.1

**Comment by bovine3dom** [comment]
See #1208 for ways to mitigate. I agree - it's really annoying, but it will take us a while to fix.

**Comment by bovine3dom** [comment]
Vim Vixen sped up their completions recently - perhaps we can learn something from it: https://github.com/ueokande/vim-vixen/pull/736/files

**Comment by Rummskartoffel** [comment]
I think a big source of lag with completions is the way they are displayed. For every completion option displayed, the HTML elements are created anew using the HTML tagged template (`` html`<tr ...>` ``). I tried just naïvely doing that part by hand in the excmd completions, and the performance profile looks a lot better already.

**Comment by bovine3dom** [comment]
Yeah the plan was to move to some sort of framework that made making / updating HTML quick and "easy", so I think we sort of knew that that was the problem.

We made that plan 3 or 4 years ago though so it's unlikely to happen any time soon

**Comment by mozbugbox** [comment]
> I think a big source of lag with completions is the way they are displayed. For every completion option displayed, the HTML elements are created anew using the HTML tagged template (`` html`<tr ...>` ``). I tried just naïvely doing that part by hand in the excmd completions, and the performance profile looks a lot better already.

When attaching new result to the completion div, it seemed that the `optionContainer` node are not detached from the live Dom. This could slow down the renderer when large amount  mutation happened.

https://github.com/tridactyl/tridactyl/blob/fcdcdc028f01d61589d3e84923f4d3318967c3aa/src/completions.ts#L301-L321
